### ISSUE TITLE
BatchNormalization's avg_var initialized with zeros can cause division by zero

### ISSUE THREAD

**Issue by muupan** [chainer/chainer#879]
Currently, `BatchNormalization`'s `avg_var` is initialized with zeros. This will cause division by zero when called it with test=True before test=False, so it might be better we use another values, say, ones, to initialize `avg_var`.

Calling `BatchNormalization` with test=True before test=False can seem unnatural, but it can happen in the reinforcement learning context: an agent selects an action by evaluating its value function or policy represented by a NN, observes consequences, and then (if it has enough experiences) updates its value function or policy.

**Comment by delta2323** [comment]
@muupan Do you think we should reopen this issue?

**Comment by muupan** [comment]
I think it is good to fix this issue by initializing var with 1s. PyTorch seems doing so.
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L24

Are there any opinions against doing so?

**Comment by beam2d** [comment]
One concern is that it changes the models learned with the same script. I think the effect is negligible for most use cases (the effect is made exponentially small against the number of training iterations), but it would be better to leave a way to reproduce the current behavior.

**Comment by delta2323** [comment]
One simple solution would be to add an option such as `initialize_population_variance_with_zero` to `__init__`. Or more generally, we can give an initializer for initializing `avg_var` (and possibly `avg_mean`), although initializers are currently used for not persistent values but parameters.

**Comment by beam2d** [comment]
+1 for using initializers for `avg_var` (and `avg_mean` for consistency).

**Comment by delta2323** [comment]
@muupan How about you?

**Comment by toslunar** [comment]
I'd rather not add any options.  Just fix it, please.  Even if someone exploited the fact that `avg_var` is initialized with zeros, he or she can fix code easily by
```
bn = chainer.links.BatchNormalization(_)
bn.avg_var[...] = 0
```
(and all Chainer has to do is to write it in the version-up guide).  To anyone else, including beginners, `avg_var` should not be 0.

**Comment by beam2d** [comment]
I will work on it.
